### ISSUE TITLE
Implement C# 8 non-nullable support

### ISSUE THREAD

**Issue by dotnetjunkie** [simpleinjector/SimpleInjector#649]
C# 8 has non-nullable support. This is of great advantage of users of libraries that can now be informed by the compiler when a library returns a value that can be null.

This should be added to Simple Injector

ATM, the feature must be enabled by editing the project file. Change the project's language version to `C# 8.0 (beta)` and add the following elements after the `<LangVersion>` element in the .csproj file:

``` xml
    <NullableReferenceTypes>true</NullableReferenceTypes>
    <NullableContextOptions>enable</NullableContextOptions>
```

UPDATE:

With Visual Studio 2019 16.2, the project file needs to be changed to the following:

``` xml
    <LangVersion>preview</LangVersion>
    <Nullable>enable</Nullable>
```

**Comment by dotnetjunkie** [comment]
This will, although noble, unfortunately not be an easy task to accomplish. I did a quick scan, and found the following issues that should be fixed before starting this undertaking (I will update the list when new issues are discovered):

* **Bug**: [InjectionConsumerInfo.Equals](https://github.com/simpleinjector/SimpleInjector/blob/master/src/SimpleInjector/InjectionConsumerInfo.cs#L105) can throw a `NullReferenceException`.
* **Design**: The `PredicateContext.Consumer` property will be `null` when a conditional registrsation is resolved as root type. With C# 8.0, this forces the user to do explicit `null` checks to prevent the compiler from tripping up. Changing  `PredicateContext.Consumer` to never return null, however, is a breaking change.
* [more issues will be added here]

**Comment by dotnetjunkie** [comment]
[feature-649](https://github.com/simpleinjector/SimpleInjector/tree/feature-649) branch created.

**Comment by dotnetjunkie** [comment]
I’ve finished adding non-nullable support to the core library. Although this doesn't complete this work item, it does mean an important step has been made. It proved to be an enormous amount of work (several full days of work). It did, however, provide me with some very interesting insights into the Simple Injector code base.

The main focus of this initial work was to:

* Make the core library compile with the C# 8.0 compiler with its non-nullable feature enabled
* Provide hints in the library’s public API when values are nullable and when they are not

In my initial [change set]( https://github.com/simpleinjector/SimpleInjector/commit/0cd49ae362c909c5b7adac97c8edd0c0500833e8) I tried to minimize the internal code changes, and instead tried to mark nullable reference types as such. By limiting the amount of changes, I reduced the risk of introducing bugs, while making the change set easier to review for others, and moving later internal refactorings and future API changes to the backlog. The issues #693, #694, #695, #696, #698, #699, #700, #701, #702, #703, and #704 are all the direct result of the insights gained from this work.

Still, this work give me an interesting new way of looking at the code base and how non-nullable support can positively influence your code base’s design. The most important realization is that nullable reference types severely complicate code, more than we actually realize. Issues #693, #696, #701, #703, #704, for instance, are all null-refrence bugs that I found in the process. It does mean, though, that I sometimes need to rethink my design (which, I think, is awesome!).

Simple Injector contains many internal types that have fields or properties that will be null in certain cases, while always be present in other cases. This, however, is a very implicit model, that is, unfortunately, used holistically throughout the code base.

To give an example, Simple Injector’s decorator sub system works either with a supplied *decorator type* or with a *factory delegate* that can produce the *decorator type*. But it is always either one or the other. That part of the code base, however, always assumes the availability of one or the other, but that’s quite hard to reason about when you’re five levels deep into the call graph. It’s better to change the data structures to make this clear, and structure your code accordingly. How to do so, however, needs to be determined on a case-by-case basis.

Most prominent downside, at least at this stage, is that C# 8.0 beta’s null flow analysis is still quite lacking. This means that C# 8.0 beta quite often reports a "possible dereference of a null reference" in cases where the value is quite obviously never null. Especially when dealing with checks like `nullable?.Value == true` and working with anonymous types inside LINQ queries, the C# compile gets confused. Funny thing is, though, that when I tried to create a MVCE for these issues, I wasn't able to reproduce them. It's when the method starts to get a bit bigger, and more complex, the compiler seems to throw the towel in.

I’m sure many things will be fixed before C# 8.0 is released. But downside of this limited flow analysis is that I sometimes had to rewrite the code a bit to make it clear to C# that the values wouldn’t be null. With the former example for instance, changing it to `nullable != null && nullable.Value` (which is exactly the same) fixed the issue. With anonymous types, I had to ‘bang’ the hell out of them (by suffixing stuff with the exclamation mark `!`).

That said, this is one of the most awesome new C# features we had in years. Equally awesome as generics and LINQ syntax. It forces you to think about nullability and forces you into a cleaner, better to reason about, design. What's not to love?

I am still considering whether or not I will push this feature out before the final release of C# 8.0 or not.

To be continued.
